passage/src/main/java/id/passage/android/PassageHosted.kt

```diff
@@ -0,0 +1,177 @@
+            val bytes = randomString.toByteArray()
+            val md = MessageDigest.getInstance("SHA-256")
+            val digest = md.digest(bytes)
+            return Base64.getUrlEncoder().withoutPadding().encodeToString(digest)
+        }
+
+        internal suspend fun finishOIDC(
+            activity: Activity,
+            code: String,
+            clientSecret: String,
+            state: String,
+        ): AuthResult? {
+            val redirectUri = "$basePathOIDC/android/$packageName/callback"
+            if (PassageHosted.state != state) {
+                throw HostedAuthorizationError("State is Invalid")
+            }
+
+            var authResult: AuthResult?
+            val client = OkHttpClient()
+            val moshi =
+                Moshi
+                    .Builder()
+                    .build()
+            val jsonAdapter = moshi.adapter(OIDCResponse::class.java)
+            val mediaType = "application/json; charset=utf-8".toMediaType()
+            val requestBody = "{\"code\":\"$code\"}".toRequestBody(mediaType)
+
+            val params =
+                listOf(
+                    "grant_type" to "authorization_code",
+                    "code" to code,
+                    "client_id" to Passage.appId,
+                    "verifier" to verifier,
+                    "client_secret" to clientSecret,
+                    "redirect_uri" to redirectUri,
+                ).joinToString("&") { (key, value) ->
+                    "$key=${URLEncoder.encode(value, "UTF-8")}"
+                }
+
+            val url = "$basePathOIDC/token?$params"
+            val request =
+                Request
+                    .Builder()
+                    .url(url)
+                    .post(requestBody)
+                    .build()
+
+            withContext(Dispatchers.IO) {
+                client.newCall(request).execute().use { response ->
+                    if (!response.isSuccessful) {
+                        if (response.code == 500) {
+                            throw ServerException("Server error : ${response.code} ${response.message}", response.code)
+                        }
+                        throw ClientException("Client error : ${response.code} ${response.message}", response.code)
+                    }
+                    val responseBody = response.body?.string()
+                    if (responseBody != null) {
+                        val apiResponse = jsonAdapter.fromJson(responseBody)!!
+                        authResult =
+                            AuthResult(
+                                authToken = apiResponse.accessToken,
+                                redirectUrl = "",
+                                refreshToken = apiResponse.refreshToken,
+                                refreshTokenExpiration = null,
+                            )
+
+                        PassageTokenStore(activity = activity).setIdToken(apiResponse.idToken)
```

Comment:
A couple thoughts here:
* Currently, we only interact with token store as in instance in `Passage` class. I recommend we keep that pattern for now.
* We should only save this id token if the user has not opted out of using our token store. See `passage.handleAuthResult` for how we handle saving tokens in the Passage class [here](https://github.com/passageidentity/passage-android/blob/main/passage/src/main/java/id/passage/android/Passage.kt#L652).
```suggestion
```